nv2a: Update `inline_value`s as rendering side-effect

abaire opened the pull request:
This updates the default `inline_value` for the normal vertex attribute whenever valid values for the normal are passed, mirroring the hardware behavior observed in https://github.com/abaire/nxdk_pgraph_tests/blob/main/tests/lighting_normal_tests.cpp.

Additional tests for all attributes:
* [test](https://github.com/abaire/nxdk_pgraph_tests/blob/4e2776729ed6a3ea86b4eb15247bc9009ed7cff9/src/tests/attribute_carryover_tests.cpp#L1)
* [results](https://github.com/abaire/nxdk_pgraph_tests_golden_results/wiki/Results-Attrib_carryover)


I'm not entirely confident that this resolves all possible paths, but it does fix #601 

![Screenshot_20220110_213034](https://user-images.githubusercontent.com/448413/148886623-7d35bd2b-6f8a-4cb6-afd9-df3b6024f505.png)

(note: screenshot was from my dev branch that also includes fix for #599, without which the minimap would be opaque but the lower left panel would still be fixed)

Shoegzer commented:
With this PR, JSRF crashes immediately prior to the main menu (after the loading screen), with the following message in term:

`qemu-system-i386: ../hw/xbox/nv2a/pgraph.c:6273: pgraph_bind_vertex_attributes: Assertion attr->gl_type == GL_FLOAT && "Need to convert non-float normal."' failed.
`

abaire commented:
I'm hoping that I'm being overly cautious with that assert so it can probably be removed (though I'm interested in seeing what is actually being used if it's not a float).

Unfortunately it seems like I can no longer get past the loading screen even with a clean build from `master` so I'm putting this back into draft mode until I can verify that it's safe to remove the assert.

abaire commented:
Looks like it uses unsigned byte for normals after the loading screen. Removed the assert and fixed the handling to support arbitrary types instead of assuming they'll always be floats..I also had an error in how I was grabbing the last element, so as an added bonus the coloring is more correct now as well.

Thanks for testing @Shoegzer !

abaire commented on hw/xbox/nv2a/pgraph.c:
```diff
@@ -6433,6 +6434,26 @@ static void pgraph_bind_vertex_attributes(NV2AState *d,
+                uint32_t element_size = attr->size * attr->count;
+                const uint8_t* last_entry = d->vram_ptr + start;
+                if (stride) {
+                    last_entry += stride * (num_elements - 2);
```
Double checking on these. Is it possible that `num_elements` doesn't actually need the +1 at the top of the function (I'd expect that these should be `num_elements - 1`, but that definitely reads past the last element)? 

Not sure if there's an off-by-one error or just an error in my assumptions based on the naming.
- In my pgraph test I'm sending a single triangle (3 elements). 
- `count` is decremented by one before being sent to DRAW_ARRAYS as per spec, but that is handled [here](https://github.com/mborgerson/xemu/blob/01c2f0fa74d1f67ec135eaeed81f029058ef25a0/hw/xbox/nv2a/pgraph.c#L2982) so pg->draw_arrays_max_count ends up being correct (3)
- When the BEGIN_END is ended, the vertex attributes are bound w/ `3` as `max_element`. The naming and the +1 makes it sound like `max_element` is expected to be the maximum valid element (`2` in my test case) rather than the max count (`3`) that's being passed in.

abaire replied:
Testing with Gun Valkyrie I feel more confident that the `- 2` is wrong and the `DRAW_ARRAYS` path should be passing `max_element` rather than `count`. Added a fix for that in this PR.

Shoegzer commented:
Tested and confirmed the assert/crashing is gone, and the bottom-left panel looks good.  I haven't merged your other fix for the map itself, but with this PR alone it now shows:

![pic](https://user-images.githubusercontent.com/12475371/149054407-650f6b77-c449-4cb6-8aa4-a61d3349a3d1.png)

Nice work!

abaire commented:
Just happened to find that this crashes Gun Valkyrie, where a call is made with `num_elements` == 1

Shoegzer commented:
That's too bad.  Not sure what that means in terms of getting this PR merged, though from what you say here:

> I also had an error in how I was grabbing the last element, so as an added bonus the coloring is more correct now as well.

Would the panel color still be mostly correct if you avoided the condition for `num_elements == 1`?  If so, would it make any sense to move that part to another PR and merge this one?  Happy to help test any workarounds either way.

abaire commented on hw/xbox/nv2a/pgraph.c:
```diff
@@ -2502,7 +2502,7 @@ DEF_METHOD(NV097, SET_BEGIN_END)
 
             pgraph_bind_vertex_attributes(d, pg->draw_arrays_min_start,
-                                          pg->draw_arrays_max_count, false, 0);
+                                          pg->draw_arrays_max_count - 1, false, 0);
```
Based on the usage in `pgraph_bind_vertex_attributes`, it looks like the expectation is that this should be the last used element rather than the count.

Setting this to count requires a -2 to access the last used element in JSRF (and causes crashes in cases that use inline elements, such as Gun Valkyrie). As far as I can see, providing `max_element` in this path still works correctly in the games I've tested.

abaire commented:
> That's too bad. Not sure what that means in terms of getting this PR merged, though from what you say here:
> 
> > I also had an error in how I was grabbing the last element, so as an added bonus the coloring is more correct now as well.
> 
> Would the panel color still be mostly correct if you avoided the condition for `num_elements == 1`? If so, would it make any sense to move that part to another PR and merge this one? Happy to help test any workarounds either way.

I think the crash was due to me working around what I think is a bug elsewhere. To fix the crash I fixed the other bug, so I think this should still be merge-able. It'd be great if you could verify that it still works for you with this latest push and try out as many other games as you can to see if it causes any weird glitches. The underlying fix will affect many more things than lighting but ideally you would not see any actual difference.

Shoegzer commented:
You bet.  Tested JSRF and it still works fine.  I also tested Halo:CE, Halo 2 and those work as well as before at least.  Star Wars KOTOR regressed, however, as eyes are missing from the character select screen (though this issue occurred before until very recently):

PR:
![prpic](https://user-images.githubusercontent.com/12475371/150658023-80221288-b928-4544-9352-e3e31d73957e.png)

Master:
![masterpic](https://user-images.githubusercontent.com/12475371/150657976-63fb6e8c-1d62-4529-82bf-da85e5d0570f.png)

Side-note since you asked: I have many other discs but I don't want to dump them to non-archival quality (xiso) images for various reasons as mentioned in #238 or I'd test those for you too.  Hopefully xemu can support proper isos in the future (though I'm patient and respect priorities).

abaire commented:
Ah, that'd almost certainly mean that I'm wrong about the count versus max somehow, or that there's some interesting dependency on the draw mode that needs to be accounted for. I own KOTOR so I'll do some debugging in a few days when I'm back with my hardware.

abaire commented:
Well, the good news is that this seems to have been a simple case of the fix branch being older than the fix for KOTOR. I rebased to the latest master and the eyes now look correct. @Shoegzer would you mind re-verifying with the latest push?

The bad news is that I wrote a [much more in-depth test](https://github.com/abaire/nxdk_pgraph_tests/blob/321b73ce7de19735bd1f21c4c80329ac1e3edb48/src/tests/attribute_carryover_tests.cpp#L1) in anticipation of tracking this down and found that the behavior in the case of inline elements is slightly wrong (it should always take the last _rendered_ vertex, the code I wrote takes the last buffered vertex). I also verified that all the other attributes  was able to test are supposed to carry over in the same way the normals do (diffuse, specular, tex coords, almost certainly the rest but I can't easily test them with nxdk's Cg-based shaders).

@mborgerson would you be OK to review and merge this PR as is and I'll do a followup? 
The handling of the inline elements and additional attributes will add some complexity so I'm inclined to get this part that actually fixes a game in and then expand on it (I don't know if any other games suffer similar problems due to missing carryover).

mborgerson commented:
> The bad news is that I wrote a [much more in-depth test](https://github.com/abaire/nxdk_pgraph_tests/blob/321b73ce7de19735bd1f21c4c80329ac1e3edb48/src/tests/attribute_carryover_tests.cpp#L1) in anticipation of tracking this down and found that the behavior in the case of inline elements is slightly wrong (it should always take the last _rendered_ vertex, the code I wrote takes the last buffered vertex). I also verified that all the other attributes was able to test are supposed to carry over in the same way the normals do (diffuse, specular, tex coords, almost certainly the rest but I can't easily test them with nxdk's Cg-based shaders).

This is what I had expected. Thanks for creating a test case, it is actually good news.

> @mborgerson would you be OK to review and merge this PR as is and I'll do a followup? The handling of the inline elements and additional attributes will add some complexity so I'm inclined to get this part that actually fixes a game in and then expand on it (I don't know if any other games suffer similar problems due to missing carryover).

Let's just fix it generally, it should not be too difficult.

abaire commented:
> > The bad news is that I wrote a [much more in-depth test](https://github.com/abaire/nxdk_pgraph_tests/blob/321b73ce7de19735bd1f21c4c80329ac1e3edb48/src/tests/attribute_carryover_tests.cpp#L1) in anticipation of tracking this down and found that the behavior in the case of inline elements is slightly wrong (it should always take the last _rendered_ vertex, the code I wrote takes the last buffered vertex). I also verified that all the other attributes was able to test are supposed to carry over in the same way the normals do (diffuse, specular, tex coords, almost certainly the rest but I can't easily test them with nxdk's Cg-based shaders).
> 
> This is what I had expected. Thanks for creating a test case, it is actually good news.
> 
> > @mborgerson would you be OK to review and merge this PR as is and I'll do a followup? The handling of the inline elements and additional attributes will add some complexity so I'm inclined to get this part that actually fixes a game in and then expand on it (I don't know if any other games suffer similar problems due to missing carryover).
> 
> Let's just fix it generally, it should not be too difficult.
> 
> By the way, the arrays implementation is flawed. I have a patch for it that notably fixes some sports games, but it needs a couple of remaining improvements. You are welcome to work on it, or I'll return to it shortly

abaire commented:
Accidentally closed, sorry! This should be ready for review now.

> Let's just fix it generally, it should not be too difficult.

Sounds good. PR is updated, all testable cases now work correctly. I can't fully test the inline buffers case because xemu doesn't yet handle all of the attribute setter functions (e.g., `SET_DIFFUSE_COLOR` w/ 4 floats which my test program uses because I haven't gotten around to making use of floats versus DWORDs conditional).

> By the way, the arrays implementation is flawed. I have a patch for it that notably fixes some sports games, but it needs a couple of remaining improvements. You are welcome to work on it, or I'll return to it shortly

Cool, I can take a look, I assume it's [fix/arrays](https://github.com/mborgerson/xemu/tree/fix/arrays)? Do you know offhand which games it fixes?

While writing this PR I noticed that there's an off-by one in the call to `pgraph_bind_vertex_attributes` when handling `DRAW_ARRAYS` but I've stomped on my comments a couple times now. Mentioning it here in case you haven't read the full conversation history.

Triticum0 commented:
It is fix/array and the game it fixes are on this issue #329.

Shoegzer commented:
> Well, the good news is that this seems to have been a simple case of the fix branch being older than the fix for KOTOR. I rebased to the latest master and the eyes now look correct. @Shoegzer would you mind re-verifying with the latest push?

I was wondering if this was a rebase issue since the eyes were fixed so recently in master, glad to know it was that simple.  In any event, KOTOR eyes are fine now with your PR, and of course JSRF lower-left panel looks good too.  Nice.

mborgerson commented on hw/xbox/nv2a/pgraph.c:
```diff
@@ -2496,7 +2496,9 @@ DEF_METHOD(NV097, SET_BEGIN_END)
 
             pgraph_bind_vertex_attributes(d, pg->draw_arrays_min_start,
-                                          pg->draw_arrays_max_count, false, 0);
+                                          pg->draw_arrays_max_count - 1,
```
We should rename `draw_arrays_max_count` while we are at it

mborgerson replied:
> Do you mean we should change it to something like `draw_arrays_max_element` and remove the increment in the `DRAW_ARRAYS` handler? At the moment I believe it's tracking the count rather than the index of the last element so the current name seems to match the behavior.

`draw_arrays_max_count` tracks maximal (start + count) [max_element + 1]
 
> Or do you mean we should change the suffix to `_length` to align with the other lengths?

`draw_arrays_length` is already taken (number of arrays)

Actually nevermind, I'll just fix this in a follow up

abaire commented on hw/xbox/nv2a/pgraph.c:
```diff
@@ -6429,6 +6437,26 @@ static void pgraph_bind_vertex_attributes(NV2AState *d,
+        } else {
+            last_entry += element_size * provoking_element_index;
+        }
+        memcpy(attr->inline_value, last_entry, element_size);
```
Added explicit handling of all the cases from `SET_VERTEX_DATA_ARRAY_FORMAT` below.

mborgerson commented on hw/xbox/nv2a/pgraph.c:
```diff
@@ -6538,11 +6544,60 @@ static void pgraph_update_memory_buffer(NV2AState *d, hwaddr addr, hwaddr size,
+static void pgraph_update_inline_value(VertexAttribute *attr,
+                                       const uint8_t *data)
+{
+    if (attr->gl_type == GL_FLOAT) {
```
Let's use a `switch` and test against `attr->format` and panic on `default` case like SET_VERTEX_DATA_ARRAY_FORMAT

abaire replied:
Testing against `attr->format` feels like it introduces some duplication to me. 

E.g., if for some reason `FORMAT_TYPE_CMP` was changed to be a `GL_FOOBAR` instead of a `GL_INT` it'd silently do the wrong thing here. The intended behavior here is dependent on the combination of the `gl_type` and normalize/conversion flags and to me it doesn't seem like it necessarily needs to know or care about what the HW wants the format to be.

Changed to a switch with a panic, let me know if you feel strongly about the check against `format` (I realize my objection is a bit pedantic since I doubt there's a scenario where the gl types would actually change), but going that route feels like it'd also need an assert against `gl_type` which seems like it's effectively equivalent.

mborgerson replied:
The input to this function is of the nv2a defined type, not the gl type. If cmp was changed to some other GL type it would be irrelevant for this matter, because we are always converting to float and we should know that the input is compressed from the nv2a type. Moreover, testing against the nv2a type keeps the code more generic, less tied to GL.

abaire replied:
Makes sense, I'd erroneously assumed that there was more nv2a -> GL conversion happening to the raw data buffer than there is in practice - updated.

mborgerson commented on hw/xbox/nv2a/pgraph.c:
```diff
@@ -6538,11 +6544,61 @@ static void pgraph_update_memory_buffer(NV2AState *d, hwaddr addr, hwaddr size,
+        case NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_CMP: {
+            /* 3 signed, normalized components packed in 32-bits. (11,11,10) */
+            typedef struct {
+                signed int x : 11;
```
I agree this looks nice, but let's do a shift and mask instead of depending on C struct bitfields; since the compiler can choose how to order them, we cannot depend on it to be correct

abaire replied:
Done.

mborgerson commented on hw/xbox/nv2a/pgraph.c:
```diff
@@ -6602,6 +6659,27 @@ static void pgraph_bind_vertex_attributes(NV2AState *d,
+        // Carry over the attribute value from the provoking vertex to mimic
+        // hardware behavior.
+        uint32_t provoking_element_index = provoking_element - min_element;
+        uint32_t element_size = attr->size * attr->count;
```
Let's use `size_t` for this variable

mborgerson commented on hw/xbox/nv2a/pgraph.c:
```diff
@@ -6602,6 +6659,27 @@ static void pgraph_bind_vertex_attributes(NV2AState *d,
+            last_entry = d->vram_ptr + start;
+        } else {
+            last_entry = (uint8_t*)pg->inline_array;
+            last_entry += attr->inline_array_offset;
```
This can be a single line wlor

mborgerson commented on hw/xbox/nv2a/pgraph.c:
```diff
@@ -6538,11 +6544,68 @@ static void pgraph_update_memory_buffer(NV2AState *d, hwaddr addr, hwaddr size,
+static void pgraph_update_inline_value(VertexAttribute *attr,
+                                       const uint8_t *data)
+{
+    switch (attr->format) {
```
Let's add a check here to make sure `attr->count < 4`

mborgerson commented on hw/xbox/nv2a/pgraph.c:
```diff
@@ -6602,6 +6666,26 @@ static void pgraph_bind_vertex_attributes(NV2AState *d,
 
         glEnableVertexAttribArray(i);
+
+        // Carry over the attribute value from the provoking vertex to mimic
```
nit: The function below already states this, so this comment is redundant

mborgerson commented on hw/xbox/nv2a/pgraph.c:
```diff
@@ -6602,6 +6666,26 @@ static void pgraph_bind_vertex_attributes(NV2AState *d,
+        assert(element_size <= sizeof(attr->inline_value));
+        const uint8_t *last_entry;
+
+        if (!inline_data) {
```
nit: Let's remove the `!` and swap the blocks

mborgerson commented:
Thanks!
